sys/riscv/riscv/trap.c

```diff
@@ -138,14 +136,18 @@ cpu_fetch_syscall_args(struct thread *td)
 		if (SV_PROC_FLAG(td->td_proc, SV_CHERI))
 			stack_args = (char * __capability)td->td_frame->tf_sp;
 #endif
+	} else {
+		*dst_ap = *ap++;
+		dst_ap++;
 	}
 
-	if (sa->code >= p->p_sysent->sv_size)
+	if (__predict_false(sa->code >= p->p_sysent->sv_size))
 		sa->callp = &p->p_sysent->sv_table[0];
 	else
 		sa->callp = &p->p_sysent->sv_table[sa->code];
 
 	sa->narg = sa->callp->sy_narg;
+	KASSERT(sa->callp->sy_narg <= nitems(sa->args), ("Too many syscall arguments!"));
```

Comment:
```suggestion
	KASSERT(sa->callp->sy_narg <= nitems(sa->args), ("Syscall %d takes too many arguments!", sa->code));
```
That being said, this probably ought to be verified when the sysent is installed and when modules install syscalls rather than in this path.

Reply:
Also, in the syscall() and __syscall() cases this needs to be `nitems(sa->args) - 1`.

Reply:
Erm... why?  In case of syscall() and __syscall() the arguments are shifted in registers, but in sa->args they are in their normal order, ie the same as if the "target syscall" was called directly, without going through syscall(2).